### ISSUE TITLE
Rule blocked: CORERULES-9562 - invalid_duration operator not accurate

### ISSUE THREAD

**Issue by ASL-rmarshall** [cdisc-org/cdisc-rules-engine#828]
**File with the JSON containing the full request**: [Request.txt](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBMWJDQmc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--47ff511d6ff8784be385a95bb75852368d30b8c7/Request.txt)

**Links to related JIRA Tickets**
- https://jira.cdisc.org/projects/CORERULES/issues/CORERULES-9562
- https://jira.cdisc.org/projects/CORERULES/issues/CORERULES-9563
- https://jira.cdisc.org/projects/CORERULES/issues/CORERULES-9564

**Rule Information**
- **Standard**: USDM
- **Rule ID**: DDF00060
- **Rule Description**: The value for each timing must be specified in ISO 8601 duration format.

**Describe the bug**
The `invalid_duration` operator does not accurately report invalid ISO 8601 durations - it does not report some invalid durations as invalid and it reports other valid durations as invalid.  Ideally, this operator should be based on the full ISO 8601 specification, though it's unclear which version should be used - e.g., 2004 or 2019. However, at a minimum, I think this operator should align with the ISO 8601 duration specifications and guidance in the SDTMIG.

Examples of:
- Invalid durations not reported as invalid:
    - Mixing weeks with other time units - e.g., P1M2W, P3W1D, P1WT4H.
- Valid durations reported as invalid:
    - Decimals for "lowest order"/rightmost component (note that a zero is required before the period with decimal values less than 1) - e.g., P1Y2.5M, P3.5W, P8DT12.5H, P0.5D, PT1.25H
    - Negative durations (though this may depend on the ISO 8601 version implemented) - e.g., -P7D, -PT24H

**Error returned from Rule Engine**
```json
{
    "TIMING": [
        {
            "executionStatus": "success",
            "dataset": "Timing.xpt",
            "domain": "TIMING",
            "variables": [
                "id",
                "name",
                "parent_entity",
                "parent_id",
                "parent_rel",
                "value",
                "valueLabel"
            ],
            "message": "The value attribute of the timing is not in ISO 8601 format.",
            "errors": [
                {
                    "value": {
                        "parent_entity": "ScheduleTimeline",
                        "value": "-P2D",
                        "name": "TIM7",
                        "id": "Timing_7",
                        "parent_rel": "timings",
                        "parent_id": "ScheduleTimeline_4",
                        "valueLabel": "2 days"
                    },
                    "row": 2
                },
                {
                    "value": {
                        "parent_entity": "ScheduleTimeline",
                        "value": "P4.5W",
                        "name": "TIM8",
                        "id": "Timing_8",
                        "parent_rel": "timings",
                        "parent_id": "ScheduleTimeline_4",
                        "valueLabel": "4.5 Weeks"
                    },
                    "row": 3
                }
            ]
        }
    ]
}
```

**Expected behavior**
For this example, "-P2D" and "P4.5W" should not be reported as invalid durations (assuming negative durations are allows - see above), and "P2W1D" should be reported as invalid.

**Comment by chowsanthony** [comment]
Duration vs. elapsed time: Specially in SDTM, the --DUR variable is about duration, expressed using the ISO 8601's P notation. In this case, there can't be a negative duration. Elapsed time in the --ELTM variable, however, can contain negative values, e.g., -P15M.

**Comment by ASL-rmarshall** [comment]
@chowsanthony I'm not sure that the SDTM/SDTMIG specifies either the version of ISO 8601 that applies or that there's a difference between representations in --DUR vs --ELTM (both are defined as have "ISO 8601 duration" format).  While ISO 8601:2004 specifies duration as "non-negative quantity attributed to a time interval...", under ISO 8601:2019-2 (Date and time - Representations for information interchange - Part 2: Extensions), negative durations are defined as "duration in the reverse direction to the proceeding time scale". See section 3.1.1.7 [here](https://www.iso.org/obp/ui/#iso:std:iso:8601:-2:ed-1:v1:en).

I don't have a copy of the full specification (not wanting to part company with CHF 194), but I have seen references that suggest negative durations can be represented with either a minus sign prefix or embedded minus sign(s).

For this operator, I guess we need to define (and document) what constitutes an "invalid" duration.

**Comment by SFJohnson24** [comment]
@ASL-rmarshall what exactly will DDF need in terms of the definition duration?  Will negative durations be permissible for DDF?

**Comment by ASL-rmarshall** [comment]
@SFJohnson24 Negative durations are not allowed for the USDM attributes referenced in the 3 rules linked above. In fact, I added a specific "value contains any hyphens" `Check` clause to the rule specifications (in addition to "value is an invalid duration") to report negatives even if they're not flagged as invalid durations.

If the `invalid_duration` operator will report negative durations as invalid, then this implies that it's based on the ISO 8601:2004 specification - which is fine, as long as anyone who wants to use the operator knows what it will report as invalid.  However, I think this might limit the operator's applicability for SDTM rules.  There will definitely be SDTM rules (e.g., for --EVLINT variables) to check for invalid (ISO 8601) durations that are allowed to be negative.
